CosmosApiChangeFeedForMessageRetention/handler.ts

```diff
@@ -0,0 +1,215 @@
+          });
+          // eslint-disable-next-line no-underscore-dangle
+          return `the timestamp of the document ${document.id} (${document._ts}) is after the RELEASE_TIMESTAMP ${RELEASE_TIMESTAMP}`;
+        }
+      )
+    ),
+    TE.chain(
+      TE.fromPredicate(
+        hasNotTTL,
+        () => `the document ${document.id} has a ttl already`
+      )
+    )
+  );
+
+/**
+  Return a Right if the document is eligible for the ttl
+ */
+
+export const setTTLForMessageAndStatus = (
+  document: RetrievedMessageStatus,
+  messageStatusModel: MessageStatusModel,
+  messageModel: MessageModel
+): TE.TaskEither<never, RetrievedMessageStatus> =>
+  pipe(
+    messageModel.patch(
+      [document.messageId, document.fiscalCode as FiscalCode],
+      {
+        ttl: TTL_VALUE
+      } as Partial<MessageStatus>
+    ),
+    TE.mapLeft((err: CosmosErrors) => {
+      throw new Error(
+        `Something went wrong trying to update the message ttl | ${JSON.stringify(
+          err
+        )}`
+      );
+    }),
+    TE.chain(() =>
+      messageStatusModel.updateTTLForAllVersions(
+        [document.messageId],
+        TTL_VALUE
+      )
+    ),
+    TE.mapLeft((err: CosmosErrors) => {
+      throw new Error(
+        `Something went wrong trying to update the message-status ttl | ${JSON.stringify(
+          err
+        )}`
+      );
+    }),
+    TE.map(() => document)
+  );
+
+export const isRejectionReasonDefined = (
+  retrievedMessageStatus: RetrievedMessageStatus
+): boolean =>
+  retrievedMessageStatus.status === RejectedMessageStatusValueEnum.REJECTED &&
+  retrievedMessageStatus.rejection_reason !== undefined;
+
+/**
+  Handle the logic of setting ttl for those message-status entries related to 
+  non existing users for IO.
+ */
+
+export const handleSetTTL = (
+  messageStatusModel: MessageStatusModel,
+  messageModel: MessageModel,
+  profileModel: ProfileModel,
+  telemetryClient: TelemetryClient,
+  documents: ReadonlyArray<RetrievedMessageStatus>
+): TE.TaskEither<
+  string | never,
+  ReadonlyArray<RetrievedMessageStatus>
+  // eslint-disable-next-line max-params
+> =>
+  pipe(
+    documents,
+    RA.map((retrievedDocument: RetrievedMessageStatus) =>
+      pipe(
+        retrievedDocument,
+        isEligibleForTTL(telemetryClient),
+        TE.chainW(
+          flow(
+            // before all we check if the rejectionReason is defined
+            TE.fromPredicate(
+              isRejectionReasonDefined,
+              () => "rejectionReason not defined"
+            ),
+            TE.fold(
+              () =>
+                // the rejection reason is not defined so we need to call the profileModel in order to verify if the user exists
+                pipe(
+                  profileModel.findLastVersionByModelId([
+                    // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
+                    retrievedDocument.fiscalCode!
```

Comment:
Thi should not occurs but what id fiscalCode is undefined here?

Reply:
What if we call the `FiscalCode.decode` just before and in case of a left we simply throw an error?

Reply:
@AleDore I ended up with this [9307d87](https://github.com/pagopa/io-functions-messages-cqrs/pull/20/commits/9307d8726e6f910921ed725dfca8d32d284542b5), I'm checking the `fiscalCode` with a decode without throwing any error in case of a `Left` to avoid useless `retry`